Improved termination heuristic when response is < batchsize/2 and the…

krosenvold opened the pull request:
… resulting event stream has contigious sequence numbers.

I'm unsure if its a bug or by design, but the current sql support can/will fail if a given aggregate has jumps in sequence numbers that are > batchSize.
This feature requires that batchsize be doubled from its current value to retain existing (somewhat questionable) behaviour.

I will admit that I find the batchSize vs maxGapOffset somewhat confusing (the docs on maxGapOffset are beyond my comprehension).
I somehow perceive batchSize to be at a lower layer controlling interactions with a single store. As such
this termination heuristic is applied at the same level.

It might be that someone needs to rethink the overall architecture somewhat around this.

krosenvold commented:
This pr can hopefully be applied on top of  #1689  or straight on master. It is an additional feature and totally independent of 1689. I have tried to use all my conflict-avoidance skills to make these two prs individually mergable.

Given the current "feature" of not supporting gaps in event sequence numbers larger than batchSize, users should considering doubling batch size to retain existing behaviour.

krosenvold commented:
I see the failing testcases, which suggest to me that coupling the early termination criteria to batch size might be a bad idea. I am unsure of the required support for sequence number jumps. For good (or bad) we have renumbered all of our sequence numbers in those cases we have deleted events. I suppose it might make sense to not encode the early termination criteria as batchSize /2 but make it a separate builder parameter that only works when enabled. 

I await your feedback on this before adapting further

smcvb commented:
I have suite surprisingly been looking int this same area not too long ago.
More specifically followed by issue #1682

The mentioned issue, I would imagine a window of opportunity that the retrieved batch contained no events. 
Simply because of the use of a historic and active `EventStorageEngine`, which obviously deviate with their sequence numbers.

Whenever another storage engine is used, however, I would assume the gaps between events no to exceed the batch size, ever.
In any sense, it is definitely not intended too. 
So, would you mind elaborating for which exact storage engine implementation you would desire this adjustment?
And furthermore, for which operation on said storage engine you require this change?

Thanks in advance @krosenvold!

krosenvold commented:
We're using the plain sql mysql implementation with dual databases, just like #1689. As for my observations about failing when gaps exceed batchsize they may have been done on a slightly outdated axon version, and I see that you added lots of testcases fairly recently. So I find it reasonable to just "forget" any comments I had about missing events, especially after #1689 was resolved. The issue still remains though, that the termination criteria seem to be quite heavily slanted in favour of people with large gaps in sequence numbers - to the point of giving users with "clean" databases a noticeable performance disadvantage. I see the flag fetchUntilEmpty is implemented but seems to be fairly obviously buggy because it can fail at the loss of a single sequence number in the wrong place (and is not in use anywhere).

smcvb commented:
I am uncertain what your argumentation of a "clean" database is in this case @krosenvold.
I would wager that a single `EventStore` storing your events should not, ever, have any gaps for aggregate events.
The only case gaps would occur for aggregate stream retrieval, is when using the `SequenceEventStorageEngine`, as this introduces N potential stores which carry your info.
If you would be able to shed some light on your perspective of "what a clean event store" is, I could provide better argumentation, if neccesary.

Now, going back to the "termination heuristic", it is worth the discussion whether the `BatchingEventStorageEngine` should be the place to even do this at.
As stated in the aforementioned issue #1682, we see the following [comment](https://github.com/AxonFramework/AxonFramework/issues/1682#issuecomment-763698642) from `haraldk`:

> As mentioned in the PR #1683 discussion, I think the problem is really caused by the JdbcEventStorageEngineand/or the EventStreamSpliterator in BatchEventStorageEngine. That's what I meant by "partial".
> 
> I think the main problem is in the SQL generated that sets an upper bound on the sequence number, rather than using limit to create the correct batch size. The current implementation will still cause problems if there are somehow gaps in the sequence numbers (as the batch fetching code in the spliterator will stop if the number of results are less than the batch size).
> 
> Perhaps this should be filed as a separate issue, otherwise I suggest leaving it open until this is addressed.

I'd lean more towards this idea too. It is the SQL performed which should be different if not everything returned whenever using the `SequenceEventStorageEngine`.
Changing this to a vendor-specific SQL which limits the number of rows returned seems to me to be an easier solution to implement.
Going the vendor-specific route in Axon Framework is however taking it a step too far, as that would require us to implement any conceivable approach out there.
This is why the `JdbcEventStorageEngine` (which I assume you are using two instances off given your description) allows you to configure the SQL being used to retrieve events as well.

Now, to conclude this comment, I am not entirely sure whether this resolves the problem you are facing.
Mainly, because I am still unsure whether you use the `JdbcEventStorageEngine`, `JpaEventStorageEngine`, `MongoEventStorageEngine` and/or the `SequenceEventStorageEngine` at all.

krosenvold commented:
Just to be certain we're discussing the same thing; I'm talking about the "sequenceNumber" column that is local to a specific aggregate. This *should* always be sequential and without gaps, even with multiple databases. (So should not be confused with the "id" column, which has gaps)

I believe the root cause of this whole "design" is that the world is a harsh and unfair place to live in. So while events in the database should not be deleted, reality has a way of challenging this every now and then. Say some runaway process created 50 million events on a single aggregate overnight. For most users this means deleting events in the database just to restore normal operations. There are also a few other "plausible" reasons why a user might want to delete events within an aggregate.

When I say our database is "clean" I simply mean that when we have deleted events, all the subsequent sequence numbers have been renumbered so the sequence numbers are contiguous.

And all of the termination logic within the sql driver indicate a certain tolerance for missing sequence numbers. There appears to be a certain consensus that this "cannot happen". but it happens - and not just to us (I am the original author of the axon plain sql driver and this tolerance has been there since before I implemented this driver!).

smcvb commented on eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/BatchingEventStorageEngine.java:
```diff
@@ -225,6 +225,8 @@ protected void validate() throws AxonConfigurationException {
 
```
I am missing a test case for this change. Would you mind adding one @krosenvold?

smcvb commented on eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/BatchingEventStorageEngine.java:
```diff
@@ -237,6 +239,7 @@ private EventStreamSpliterator(Function<T, List<? extends T>> fetchFunction,
             super(Long.MAX_VALUE, NONNULL | ORDERED | DISTINCT | CONCURRENT);
             this.fetchFunction = fetchFunction;
             this.batchSize = batchSize;
+            this.smallResultTreshold = this.batchSize/2;
```
What do you think about moving the setting of the `smallResultTreshold` to an actual method so that we have a little more control over it?

If I've followed the current implementation correctly enough, this feature is in essence replacing the `BatchingEventStorageEngine#fetchForAggregateUntilEmpty()` method.
This method was introduced to help people that do not have clean event stores (unlike yourself).
Hence why the field it sets is validated at the start of the `tryAdvance` method.

Having such a method would allow the batching implementation to be ambiguous about the outcome. 
In contrast, the JDBC implementation (for which you are making this adjustment) could override the result to half the batch size.
Added, I think this method should reside in the builder of the `BatchingEventStorageEngine`. 
This would greatly ease the configuration for users, as they no longer have to provide a custom implementation to override the method.

What do you think, @krosenvold?

smcvb replied:
This would, by the way, mean that the `fetchForAggregateUntilEmpty` method should be deprecated in favor of the batch size.

From there, we can adjust the check to validate if the `smallResultTreshold` equals the `batchSize`. 
If it does, we're dealing with JPA and shouldn't impose this change.
If it does not equal the `batchSize`, we can assume JDBC''s override is in place (or the user's configuration, of course).
That would warrant the adjustment you've made to optimize the heuristics.

smcvb commented on eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/BatchingEventStorageEngine.java:
```diff
@@ -247,9 +250,19 @@ public boolean tryAdvance(Consumer<? super T> action) {
                 List<? extends T> items = fetchFunction.apply(lastItem);
+                int numItems = items.size();
+                if (numItems > 0 && numItems < smallResultTreshold) {
+                    DomainEventData<T> firstFetchedItem = (DomainEventData<T>)items.get(0);
```
From the spacing of brackets and operations, I am guessing you're not using the current style guide for this change.
Would you mind setting it for this PR? You can find it [here](https://github.com/AxonFramework/AxonFramework/blob/master/axon_code_style.xml) by the way.

smcvb reviewed (changes requested):
First and foremost, our apologies for the late review here @krosenvold. 
It wasn't our intention to leave you hanging like this. 
I hope you still want to put in the time and effort to carry this pull request a little further.

To the topic at hand, though. 
I have a couple of suggestions to adjust your pull request to more closely align with the fact we're dealing with a JDBC-minded solution that's now automatically imposed on JPA users as well. 
Let me know what you think about the suggestions so that we can proceed with the discussion.

Next to the recommendation, I am missing a test case for this change, though. 
Adding one would be very beneficial.

abuijze commented:
Hi @krosenvold, not our biggest display of speedy reactions 😉, but it took a while to get my head around what you were exactly trying to achieve and how best to do that.
I have made some changes to your to make the termination heuristic configuration through the builder. That way, besides the "either empty or not full" approach that was already available by overriding a method, there is now a less intrusive way to configure the exact heuristic you'd like to use.
Let me know if this doesn't address the problem you have. Happy to discuss any changes.
